### PR TITLE
DBZ-1720 Move CI to GH Actions

### PR DESCRIPTION
**Pull request by ani-sha**
@Naros need to configure a `secrets` for `WEBHOOK_URL` for Gitter.

### CODE REVIEW

**Comment by jpechane** [comment]
@ani-sha Hi, this looks very nice! I have just few comments for improvements

- make sure that the order in `paths` field is same for all workflows, first the changed module and then dependencies. Ideally the dependencies will be graphically separated for example via comment. IIUC the deps block should be the same for all workflows
- would it be possible to use `matrix` for jobs that executes more than once per connector?
- would it be possible to add one more workflow that will trigger a new docs build when documentation is changed?

Great work!

**Comment by ani-sha** [comment]
Thanks, @jpechane for the suggestions. 
- In which case would we be needing to run the connectors more than once? I believe depending on the no of keys in a matrix, that many no of times a job will be executed.
- Surely would be adding a workflow for docs. Also had a discussion with @Naros; thought of implementing this after fixing GH actions for website.

**Comment by jpechane** [comment]
@ani-sha See for example MongoDB connector - it is keyed by `version.mongo.server` Maven property

**Comment by ani-sha** [comment]
> @ani-sha See for example MongoDB connector - it is keyed by `version.mongo.server` Maven property

yep, I can create a matrix depending on the versions for `mongodb`. Anything for `postgres`?

**Comment by jpechane** [comment]
@ani-sha For postgres there is `version.postgres.server`. Unfortunately profile names are changed as well so it needs to be somehow woven together.

**Comment by ani-sha** [comment]
@jpechane I tried using a matrix for mongodb locally, but it fails the build with this error. 
```
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Debezium Parent POM 1.4.0-SNAPSHOT:
[INFO] 
[INFO] Debezium Checkstyle Rules .......................... FAILURE [  2.882 s]
[INFO] Debezium IDE Formatting Rules ...................... SKIPPED
[INFO] Debezium Revapi Rules .............................. SKIPPED
[INFO] Debezium Parent POM ................................ SKIPPED
[INFO] Debezium API ....................................... SKIPPED
[INFO] Debezium Core ...................................... SKIPPED
[INFO] Debezium Assembly Descriptors ...................... SKIPPED
[INFO] Debezium Embedded .................................. SKIPPED
[INFO] Debezium Connector for MongoDB ..................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  58.448 s
[INFO] Finished at: 2020-11-09T09:46:02Z
[INFO] ------------------------------------------------------------------------
Error:  Unknown lifecycle phase "4.2". You must specify a valid lifecycle phase or a goal in the format <plugin-prefix>:<goal> or <plugin-group-id>:<plugin-artifact-id>[:<plugin-version>]:<goal>. Available lifecycle phases are: validate, initialize, generate-sources, process-sources, generate-resources, process-resources, compile, process-classes, generate-test-sources, process-test-sources, generate-test-resources, process-test-resources, test-compile, process-test-classes, test, prepare-package, package, pre-integration-test, integration-test, post-integration-test, verify, install, deploy, pre-clean, clean, post-clean, pre-site, site, post-site, site-deploy. -> [Help 1]
Error:  
Error:  To see the full stack trace of the errors, re-run Maven with the -e switch.
Error:  Re-run Maven using the -X switch to enable full debug logging.
Error:  
Error:  For more information about the errors and possible solutions, please read the following articles:
Error:  [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/LifecyclePhaseNotFoundException
Error: Process completed with exit code 1.
```

**Comment by jpechane** [comment]
@ani-sha Could you please show the maven command you use?

**Comment by ani-sha** [comment]
> @ani-sha Could you please show the maven command you use?

```
mvn clean install -B -pl debezium-connector-mongodb -am -Passembly -Dcheckstyle.skip=true -Dformat.skip=true -Drevapi.skip -Dversion.mongo.server= ${{ matrix.version-mongo-server }} -Dorg.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener=warn
```

**Comment by jpechane** [comment]
@ani-sha Try removing the space between `=` and `${` in `-Dversion.mongo.server= ${{ matrix.version-mongo-server }}`

**Comment by ani-sha** [comment]
> @ani-sha Try removing the space between `=` and `${` in `-Dversion.mongo.server= ${{ matrix.version-mongo-server }}`

@jpechane Overcoming the first error throws a new error.

```
[INFO] Tests run: 113, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] 
[INFO] --- maven-jar-plugin:3.0.2:jar (default-jar) @ debezium-connector-mongodb ---
[INFO] Building jar: /home/runner/work/debezium/debezium/debezium-connector-mongodb/target/debezium-connector-mongodb-1.4.0-SNAPSHOT.jar
[INFO] 
[INFO] --- maven-source-plugin:3.1.0:test-jar-no-fork (attach-test-sources) @ debezium-connector-mongodb ---
[INFO] Building jar: /home/runner/work/debezium/debezium/debezium-connector-mongodb/target/debezium-connector-mongodb-1.4.0-SNAPSHOT-test-sources.jar
[INFO] 
[INFO] --- maven-jar-plugin:3.0.2:test-jar (test-jar) @ debezium-connector-mongodb ---
[INFO] Building jar: /home/runner/work/debezium/debezium/debezium-connector-mongodb/target/debezium-connector-mongodb-1.4.0-SNAPSHOT-tests.jar
[INFO] 
[INFO] --- maven-assembly-plugin:3.1.1:single (default) @ debezium-connector-mongodb ---
[INFO] Building tar: /home/runner/work/debezium/debezium/debezium-connector-mongodb/target/debezium-connector-mongodb-1.4.0-SNAPSHOT-plugin.tar.gz
[INFO] Building zip: /home/runner/work/debezium/debezium/debezium-connector-mongodb/target/debezium-connector-mongodb-1.4.0-SNAPSHOT-plugin.zip
[INFO] 
[INFO] --- docker-maven-plugin:0.31.0:build (start) @ debezium-connector-mongodb ---
[INFO] 
[INFO] --- docker-maven-plugin:0.31.0:start (start) @ debezium-connector-mongodb ---
[INFO] DOCKER> Pulling from library/mongo
[INFO] DOCKER> Digest: sha256:efc408845bc917d0b7fd97a8590e9c8d3c314f58cee651bd3030c9cf2ce9032d
[INFO] DOCKER> Status: Downloaded newer image for mongo:4
[INFO] DOCKER> Pulled mongo:4 in 8 seconds 
Error:  DOCKER> Error occurred during container startup, shutting down...
Error:  DOCKER> I/O Error [Unable to pull 'debezium/mongo-initiator:4' : {"message":"manifest for debezium/mongo-initiator:4 not found: manifest unknown: manifest unknown"} (Not Found: 404)]
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Debezium Parent POM 1.4.0-SNAPSHOT:
[INFO] 
[INFO] Debezium Checkstyle Rules .......................... SUCCESS [01:31 min]
[INFO] Debezium IDE Formatting Rules ...................... SUCCESS [  0.778 s]
[INFO] Debezium Revapi Rules .............................. SUCCESS [  0.084 s]
[INFO] Debezium Parent POM ................................ SUCCESS [01:10 min]
[INFO] Debezium API ....................................... SUCCESS [ 31.257 s]
[INFO] Debezium Core ...................................... SUCCESS [01:06 min]
[INFO] Debezium Assembly Descriptors ...................... SUCCESS [  0.057 s]
[INFO] Debezium Embedded .................................. SUCCESS [ 30.882 s]
[INFO] Debezium Connector for MongoDB ..................... FAILURE [ 37.975 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  06:31 min
[INFO] Finished at: 2020-11-09T10:13:45Z
[INFO] ------------------------------------------------------------------------
Error:  Failed to execute goal io.fabric8:docker-maven-plugin:0.31.0:start (start) on project debezium-connector-mongodb: I/O Error: Unable to pull 'debezium/mongo-initiator:4' : {"message":"manifest for debezium/mongo-initiator:4 not found: manifest unknown: manifest unknown"} (Not Found: 404) -> [Help 1]
Error:  
Error:  To see the full stack trace of the errors, re-run Maven with the -e switch.
Error:  Re-run Maven using the -X switch to enable full debug logging.
Error:  
Error:  For more information about the errors and possible solutions, please read the following articles:
Error:  [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
Error:  
Error:  After correcting the problems, you can resume the build with the command
Error:    mvn <args> -rf :debezium-connector-mongodb
Error: Process completed with exit code 1.
```

**Comment by jpechane** [comment]
@ani-sha It seems that `4.0` was converted to `4` only when passed to the mvn command

**Comment by ani-sha** [comment]
> @ani-sha It seems that `4.0` was converted to `4` only when passed to the mvn command

Yes. But `4.0` was provided in the matrix it took `4`.

**Comment by ani-sha** [comment]
@jpechane Fixed the matrix for `mongodb`.

**Comment by jpechane** [comment]
@ani-sha Nice! Do you think you'll manage to do it for postgres as well? It might be a bit more complicated as two things are updated

**Comment by ani-sha** [comment]
> @ani-sha Nice! Do you think you'll manage to do it for postgres as well? It might be a bit more complicated as two things are updated

Well the first issue I am facing right now is with the dependencies in the `postgres-connector`. So I am not able to run or test anything locally for postgres.
Screenshot from 2020-11-10 16-10-37

**Comment by jpechane** [comment]
@ani-sha Could you please share the full error message in the log? This should work the same way as MongoDB does.

**Comment by ani-sha** [comment]
> @ani-sha Could you please share the full error message in the log? This should work the same way as MongoDB does.


```
Error:  Errors: 
Error:    PostgresConnectorIT.shouldResumeStreamingFromSlotPositionForCustomSnapshot:1524->waitForSnapshotToBeCompleted:2378->AbstractConnectorTest.waitForSnapshotToBeCompleted:1035 » ConditionTimeout
[INFO] 
Error:  Tests run: 219, Failures: 0, Errors: 1, Skipped: 3
[INFO] 
[INFO] 
[INFO] --- docker-maven-plugin:0.31.0:stop (stop) @ debezium-connector-postgres ---
07:14:53.469 postgresLOG:  received smart shutdown request
07:14:53.469 postgresLOG:  autovacuum launcher shutting down
07:14:53.469 postgresFATAL:  terminating autovacuum process due to administrator command
07:14:53.759 postgresLOG:  shutting down
07:14:53.829 postgresLOG:  database system is shut down
[INFO] DOCKER> [debezium/postgres-server-test-database:latest]: Stop and removed container 0f9e4630a416 after 0 ms
[INFO] 
[INFO] --- maven-source-plugin:3.1.0:jar-no-fork (attach-sources) @ debezium-connector-postgres ---
[INFO] Building jar: /home/runner/work/debezium/debezium/debezium-connector-postgres/target/debezium-connector-postgres-1.4.0-SNAPSHOT-sources.jar
[INFO] 
[INFO] --- maven-checkstyle-plugin:3.1.1:checkstyle (check-style) @ debezium-connector-postgres ---
[INFO] Starting audit...
Audit done.
[INFO] 
[INFO] --- maven-failsafe-plugin:3.0.0-M3:verify (verify) @ debezium-connector-postgres ---
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Debezium Parent POM 1.4.0-SNAPSHOT:
[INFO] 
[INFO] Debezium Checkstyle Rules .......................... SUCCESS [  2.313 s]
[INFO] Debezium IDE Formatting Rules ...................... SUCCESS [  0.282 s]
[INFO] Debezium Revapi Rules .............................. SUCCESS [  0.069 s]
[INFO] Debezium Parent POM ................................ SUCCESS [  1.406 s]
[INFO] Debezium API ....................................... SUCCESS [  5.895 s]
[INFO] Debezium Core ...................................... SUCCESS [01:22 min]
[INFO] Debezium Assembly Descriptors ...................... SUCCESS [  0.107 s]
[INFO] Debezium Embedded .................................. SUCCESS [ 14.299 s]
[INFO] Debezium Connector for PostgreSQL .................. FAILURE [12:15 min]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  14:02 min
[INFO] Finished at: 2020-11-11T07:14:55Z
[INFO] ------------------------------------------------------------------------
Error:  Failed to execute goal org.apache.maven.plugins:maven-failsafe-plugin:3.0.0-M3:verify (verify) on project debezium-connector-postgres: There are test failures.
Error:  
Error:  Please refer to /home/runner/work/debezium/debezium/debezium-connector-postgres/target/failsafe-reports for the individual test results.
Error:  Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
```

**Comment by jpechane** [comment]
@ani-sha This looks like an intermittent failure issue so you can ignore it for now. But if it is reliable reproducibale on your machine then we are going to borrow it to find the root cause :-)

**Comment by ani-sha** [comment]
> @ani-sha This looks like an intermittent failure issue so you can ignore it for now. But if it is reliable reproducibale on your machine then we are going to borrow it to find the root cause :-)

ok sure; for postrges we would be using two matrixes one being the `version.postgres.server` with `[9.6, 10]` and other being `plugin` matrix would contain?

**Comment by jpechane** [comment]
It seems to me that `-Dversion.postgres.server=9.6-devel` could be removed and matrix will be made out of different profile settings

- assembly
- assembly,wal2json
- assembly,postgres-10,pgoutput

and the strings from above will be passed as -P{...} to the maven command

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+          mvn -version
+      - name: Cache local Maven repository
+        uses: actions/cache@v2
+        with:
+          path: ~/.m2/repository
+          key: ${{ runner.os }}-maven-${{ hashFiles('**/pom.xml') }}
+          restore-keys: |
+            ${{ runner.os }}-maven-
+
+#  web-hooks:
```
Comment (jpechane) [inline .github/workflows/build.yml:29]: I think you can remove webhooks in all workflows. There is already and integration between Github and Gitter that sends PR notifications to the dev channel. The webhooks in Travis config file were used to send notifications about build results. AFAIK there is no such link between Githuba Actions and Gitter yet.

The list of available integrations is available at https://gitlab.com/gitlab-org/gitter/services/-/tree/master/lib

So a viable solution would be sending a message in for example Jenkins API format to Gitter's Jenkins integration endpoint.

**Comment by jpechane** [comment]
@ani-sha I think there is no need to delay this later - looks good and let's give it a try!

**Comment by ani-sha** [comment]
> @ani-sha I think there is no need to delay this later - looks good and let's give it a try!
Absolutely! 🚀

**Pull request merged**